### PR TITLE
py-nbconvert: avoid install-time downloads

### PR DESCRIPTION
**Pull request by AlexanderRichert-NOAA**
This PR disables install-time downloading of css files for py-nbconvert, allowing it to be installed on non-internet connected systems. This includes adding the latest version, 7.4.1, where the css files can be prefetched as Spack resources. Tested on personal computer and Acorn (NWS) compute node.

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 1]
```diff
@@ -13,8 +13,6 @@ class PyNbconvert(PythonPackage):
     homepage = "https://github.com/jupyter/nbconvert"
     pypi = "nbconvert/nbconvert-6.0.1.tar.gz"
 
-    license("BSD-3-Clause")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py:16]: Can you undo the license changes?

Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/py-nbconvert/package.py:16]: done

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 2]
```diff
@@ -15,6 +15,7 @@ class PyNbconvert(PythonPackage):
 
     license("BSD-3-Clause")
 
+    version("7.14.1", sha256="20cba10e0448dc76b3bebfe1adf923663e3b98338daf77b97b42511ef5a88618")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py:18]: Dependencies need to be updated for this new version:
```toml
[build-system]
requires = ["hatchling >=1.5"]

requires-python = ">=3.8"
dependencies = [
    "beautifulsoup4",
    "bleach!=5.0.0",
    "defusedxml",
    "importlib_metadata>=3.6;python_version<\"3.10\"",
    "jinja2>=3.0",
    "jupyter_core>=4.7",
    "jupyterlab_pygments",
    "MarkupSafe>=2.0",
    "mistune>=2.0.3,<4",
    "nbclient>=0.5.0",
    "nbformat>=5.7",
    "packaging",
    "pandocfilters>=1.4.1",
    "pygments>=2.4.1",
    "tinycss2",
    "traitlets>=5.1",
]
```

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 3]
```diff
@@ -75,13 +76,66 @@ class PyNbconvert(PythonPackage):
...
+    @when("@:7.4.0")
     def patch(self):
         # We bundle this with the spack package so that the installer
         # doesn't try to download it.
         install(
             join_path(self.package_dir, "style.min.css"),
             join_path("nbconvert", "resources", "style.min.css"),
         )
+        # Avoid install-time downloads of css files.
+        filter_file(r"css = urlopen\(url\).read\(\)", "assert False", "hatch_build.py")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]:
```suggestion
        filter_file("css = urlopen(url).read()", "assert False", "hatch_build.py", string=True)
```

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 4]
```diff
@@ -75,13 +76,66 @@ class PyNbconvert(PythonPackage):
...
+    @when("@7.4.1:")
+    def install_css(self):
+        css = {
+            # target filename: [subdir, source filename]
+            "index.css": ["lab", "index.css"],
+            "theme-light.css": ["lab", "variables.css"],
+            "theme-dark.css": ["lab", "variables.css"],
+            "style.css": ["classic", "style.min.css"],
+        }
+        for f in css.keys():
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]: Could do:
```suggestion
        for target, (subdir, source) in css.items():
```
if you want to make it easier to read

**Comment by AlexanderRichert-NOAA** [comment]
@spackbot fix style

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 5]
```diff
@@ -75,13 +80,69 @@ class PyNbconvert(PythonPackage):
     depends_on("py-testpath", when="@5:6.4", type=("build", "run"))
     depends_on("py-lxml", when="@6.5.1:7.0", type=("build", "run"))
 
+    conflicts("^bleach@5.0.0", when="@7.14:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]:
```suggestion
    conflicts("^bleach@5.0.0")
```
This is likely true for all versions, but wasn't noticed until bleach 5.0.0 was released

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 6]
```diff
@@ -75,13 +80,69 @@ class PyNbconvert(PythonPackage):
...
+            )
+
+    @when("@:7.4.0")
     def patch(self):
         # We bundle this with the spack package so that the installer
         # doesn't try to download it.
         install(
             join_path(self.package_dir, "style.min.css"),
             join_path("nbconvert", "resources", "style.min.css"),
         )
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py]: Is it possible to replace this with a resource too? Then we don't have to worry about copyrights from copying code files.

Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/py-nbconvert/package.py]: Done. I took out the style.min.css, and resources & patching are now applied in the appropriate version range too.

File: var/spack/repos/builtin/packages/py-nbconvert/package.py [hunk 7]
```diff
@@ -75,13 +80,63 @@ class PyNbconvert(PythonPackage):
...
+
+    @run_before("install")
+    @when("@6:")
+    def install_css(self):
+        css = {
+            # target filename: [subdir, source filename]
+            "index.css": ["lab", "index.css"],
+            "theme-light.css": ["lab", "variables.css"],
+            "theme-dark.css": ["lab", "variables.css"],
+            "style.css": ["classic", "style.min.css"],
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbconvert/package.py:146]: Should this one be for all versions, not just 6+?

Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/py-nbconvert/package.py:146]: As far as I can tell it was not been getting used in the older versions.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-nbconvert/package.py:146]: @tgamblin added this way back in #3587, I would be surprised if it wasn't being used in any of those versions.

Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/py-nbconvert/package.py:146]: This was based on some test installations, but I'm happy to extend the version range.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-nbconvert/package.py:146]: Could also simply deprecate nbconvert 5 and older, I doubt anyone is still using it.

Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/py-nbconvert/package.py:146]: That works. I marked them deprecated but I'm happy to just delete them.

**Comment by AlexanderRichert-NOAA** [comment]
@spackbot fix style

File: var/spack/repos/builtin/packages/py-nbconvert/package.py (hunk 7 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-nbconvert/package.py:146]: Nah, we should deprecate first so there are no surprises.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**